src/vos/vos_iterator.c

```diff
@@ -749,50 +819,42 @@ vos_iterate_internal(vos_iter_param_t *param, vos_iter_type_t type,
 
 			reset_anchors(iter_ent.ie_child_type, anchors);
 
-			/** The child iterator may yield during iteration.
-			 *  It isn't necessary to reprobe the parent in order
-			 *  to continue iterating on the child because it
-			 *  would remain in the same location in memory.
-			 *  However, it is necessary to reprobe this iterator
-			 *  before continuing.
-			 */
-			if (need_reprobe(type, anchors)) {
-				D_ASSERT(!daos_anchor_is_zero(anchor) &&
-					 !daos_anchor_is_eof(anchor));
-				goto probe;
-			}
+			rc = advance_stage(type, 0, param, anchors, anchor, &stage,
+					   VOS_ITER_STAGE_POST, &probe_flags);
+			JUMP_TO_STAGE(rc, next, probe, out);
 		}
 
+		if (stage == VOS_ITER_STAGE_RECURSE)
+			stage = VOS_ITER_STAGE_POST;
+
+		D_ASSERT(stage == VOS_ITER_STAGE_POST);
+
 		if (post_cb) {
 			acts = 0;
 			rc = vos_iter_cb(post_cb, ih, &iter_ent, type, param, arg, &acts);
 			if (rc != 0)
 				break;
 
-			set_reprobe(type, acts, anchors, param->ip_flags);
+			/** Make sure we advance to next entry on re-probe */
+			if ((acts & (VOS_ITER_CB_SKIP | VOS_ITER_CB_DELETE)) == 0)
+				acts |= VOS_ITER_CB_SKIP;
 
-			if (acts & VOS_ITER_CB_ABORT)
-				break;
-
-			if (acts & VOS_ITER_CB_RESTART) {
-				daos_anchor_set_zero(anchor);
-				probe_anchor = NULL;
-				goto probe;
-			}
-
-			if (need_reprobe(type, anchors)) {
-				D_ASSERT(!daos_anchor_is_zero(anchor) &&
-					 !daos_anchor_is_eof(anchor));
-				goto probe;
-			}
+			rc = advance_stage(type, acts, param, anchors, anchor,
+					   &stage, VOS_ITER_STAGE_FILTER, &probe_flags);
+			JUMP_TO_STAGE(rc, next, probe, out);
 		}
-
-		rc = vos_iter_next(ih);
-		if (rc) {
+next:
+		stage = VOS_ITER_STAGE_FILTER;
```

Comment:
Honestly, I spent quite some time to understand the iteration state machine. Not sure whether others will be confused if wants to add/adjust something in the future...